### DISCUSSION TITLE
How to update to decorator stage 3？

### DISCUSSION THREAD

**Discussion by ruojianll** [facing-dev/vue-facing-decorator#44] [General]
Decorator proposal is stage 3 now. In this stage, decorator must return a class value (not any value). In v-f-d, `Component` decorator returns vue option component( a plain JS object). Let's discuss how to update to stage 3 decorator proposal.

https://github.com/tc39/proposal-decorators

I don't suggest use vue internal `__vccOpt` api, because it is not stable and not documented, so there may should be a `toOption` method to convert decorated class to vue option component.

https://github.com/vuejs/core/blob/98f1934811d8c8774cd01d18fa36ea3ec68a0a54/packages/runtime-core/src/vnode.ts#L543

```
@Decorator
class MyComp extends Vue{}

export default toOption(MyComp)
```

**Comment by ruojianll** [comment]
@armano2

**Reply by armano2** [reply]
i don't understand why do we need `toNative` that should be done in decorator itself...

-------------------

Decorators in essence are already wrappers with fancy syntax

```
export default Decorator(
  class MyComp extends Vue {}
)
```

**Reply by armano2** [reply]
just a small tip, instead of eagely releasing changes its a good practice to make PR's and ask ppl for feedback :)

its not a big deal for smaller repos, but bigger you become more important it is

**Reply by ruojianll** [reply]
@armano2 Because in stage 3, class decorator couldn't return a non-constructor value(a vue options component, also a plain js object) to instead the class itself. There is a gap between decorated class component and vue options component. So far, `toNative` is not necessary if user don't want to use it because vue internal api `__vccOpts` is still work.
I asked people in here but most of them don't care.

**Comment by rdhelms** [comment]
@ruojianll Since Vue core already has a `defineComponent` function for the Options API, maybe vue-facing-decorator could export a `defineClassComponent` function that receives the decorator's class value and does the final conversion into the necessary options API default export?

Something like
```vue
<script lang="ts">
    import { Component, defineClassComponent } from 'vue-facing-decorator'

    @Component
    class MyComponent extends Vue {}

    export default defineClassComponent(MyComponent)
</script>
```

Where `defineClassComponent` is something like
```ts
function defineClassComponent (ComponentClass) {
    const componentInstance = new ComponentClass()

    // Map any static and instance ComponentClass values to object for Options API
    const componentOptions = ...

    return defineComponent(componentOptions)
}
```

**Comment by ruojianll** [comment]
I released beta version to migrate to decorator stage 3. If you want to try, see https://facing-dev.github.io/vue-facing-decorator/#/en/next-version/next-version

**Reply by ruojianll** [reply]
@rdhelms  New api is `toNative`. And v-f-d could find which stage of decorator is working and compatible with them. The new version  couldn't work in a concurrence environment.

**Reply by rdhelms** [reply]
@ruojianll Great work on this!
I'm worried the `toNative` name will be confusing because "native" usually makes people think of writing for native mobile support (like [React Native](https://reactnative.dev/), [Web Native](https://webnative.tech/)), which has nothing to do with this work.

Could we again consider either `defineClassComponent`, or even having v-f-d export its own version of `defineComponent`? The pattern of calling `defineComponent` in the regular Vue options API is so familiar that I think people would understand it right away. It may even help people migrate to v-f-d if they decide to move from Options API to class components.

**Reply by rdhelms** [reply]
@ruojianll Also, what timeline are you thinking for this beta release? Personally, I'd be happy if we kept it beta for a long time, because the current 2.x version already seems great, and switching to the 3.x version would be an extra migration step for each of our many components. In general I think people will prefer to migrate from vue-class-components to v-f-d@2 instead of v3.

And TS has said that they will continue supporting the `experimentalDecorators` flag "[for the foreseeable future](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#differences-with-experimental-legacy-decorators)", so it doesn't seem like we're in any rush.

**Reply by ruojianll** [reply]
@rdhelms 
`defineClassComponent ` or other we can discuss in the future. In my opinion, if vue update stage3 decorator as default project setting, we should release v3. v3 is compatible with stage2 and stage3. `toNative` is not necessary, because `__vccOpt` api is still work now, the migration is almost transparent if users don't want to use `toNative`.

**Reply by ruojianll** [reply]
About the function name, 'define class component' has done when user declare the class. The function is aim to 'transform defined class component to vue options api'. I agree avoid the relationship between `toNative` and native mobile support, would you give me another one?

**Reply by rdhelms** [reply]
I think `defineClassComponent` would still be "defining" the component as much as vue core's [defineComponent](https://vuejs.org/api/general.html#definecomponent) is? To me, the "define" action is really being done by the Vue compiler when it processes the user's script block.

Vue core's `defineComponent` is [basically a no-op](https://github.com/vuejs/core/blob/9a09e476679040fcec3257d42cd0dab29e95b9f2/packages/runtime-core/src/apiDefineComponent.ts#L240), but it's still being used to help the vue SFC compiler fully define the component.

**Reply by ruojianll** [reply]
For users, use `defineComponent` to define a component in native vue.
```ts
export default defineComponent({
    // define component details here
})
```
in v-f-d, defining a component is declaring a class
```ts
@Component //defining a class component
class Comp extends Vue{

}

export default toNative(Comp) //tranform(unbox) the component to vue native component
```
This is what users think about defining a component, in my opinion.

**Reply by rdhelms** [reply]
Yeah, I definitely can see what you mean. It would be great to get some more voices in here to weigh in...I can see it both ways. Personally `defineClassComponent` is still more intuitive to me, but I could see others seeing it differently, as you've said.

I'm having a hard time coming up with any third name option 🤔

**Reply by ruojianll** [reply]
Or .. `toVanilla`? Don't hurry, there is a long time to finally release v3. 😊

**Reply by rdhelms** [reply]
Another naming idea...what if:
1. we stopped wrapping `buildComponent`'s return in `defineComponent` (so `return defineComponent(option)` would become `return option`)
2. vue-facing-decorator could then export its own `defineComponent` function (or we could go back to `defineClassComponent` if we want to stay separate from Vue's `defineComponent`), which could accept a `VueCons` type and use Vue's `defineComponent` in its return: `return defineComponent(slot.cachedVueComponent)`

That way, it can make it more clear that the "defining a component" step is still happening at the end, just like in normal Vue. The `@Component` decorator is a convenience for defining the component options, but the real "define" step (as Vue core uses the term) really is still happening at the end

**Reply by rdhelms** [reply]
@ruojianll Any more thoughts on naming re: above^ ?

I think for our own team anyway if 3.x sticks with `toNative`, we'll probably just alias it as `defineClassComponent` in our own app 🙂 That's what seems to make the most sense in our team

**Comment by rdhelms** [comment]
@ruojianll That's a good point about v-f-d needing to follow vue's decorator settings. I think you're right. So while vue stays with `experimentalDecorators` we should keep v3 in beta, but if vue removes `experimentalDecorators` then v-f-d v3 will be released?

**Reply by ruojianll** [reply]
v3 is compatible with stage3 and stage2 both. If v3 is stable enough, we could release it. It is also relative to other decorator libs. If users need to use stage3 libs, they need to use v-f-d v3.

**Reply by rdhelms** [reply]
Yeah, I guess you're right...I'm just not looking forward to having to go through all our components to add the extra line at the bottom and move the `export default` 😜

**Reply by ruojianll** [reply]
In fact, exporting your class component directly (without `toNative`) works too in 3.0.0-beta.3. Because Component decorator sets vue internal api `__vccOpts`, but if vue removes it in the future, there will be a crash. You could don't use `toNative` if you don't care the risk.

**Reply by rdhelms** [reply]
That's awesome - I didn't realize that exporting the class component directly would still work in v3 💯

**Comment by wangzhiguoengineer** [comment]
@yyx990803
Hello Evan, You are the author of Vue. Can you give some suggestions?

**Reply by rdhelms** [reply]
@wangzhiguoengineer Just FYI...I wouldn't suggest trying to pull the vue core team into vue-facing-decorator discussions...the current official stance from Vue is not to use class components at all, so I doubt that anyone there will want to devote time to supporting this package (yet).

I think we would really need to build community support for this package and for class components in general before lobbying for official support from the Vue team.

(Though if I'm wrong and if the vue team decides to start officially supporting class components via this package, that would be awesome 😀)

**Reply by ruojianll** [reply]
@yyx990803 The biggest risk is if `__vccOpts` will be stable in vue 3.

**Comment by ruojianll** [comment]
@rdhelms 
Do you think it's time to merge 3.x into main version?

**Reply by rdhelms** [reply]
@ruojianll I still haven't gotten a chance to test out the 3.x version in our codebase - we're finishing up our first app's full Vue 3 migration using vue-facing-decorator. After I finished that, I was hoping to bump the vfd version to 3.x and test it out more. Would you be willing to wait a couple more weeks before merging + publishing as default?

In the meantime, what would you think about drafting some more thorough 3.x release notes? As more people start to use the package, there should probably be a good system in place to review version changes. Either in the GitHub Release info or in a Changelog file would be awesome.

**Reply by rdhelms** [reply]
I think the docs at https://facing-dev.github.io/vue-facing-decorator/#/en/next-version/next-version are great - but I think it's common to also have a summary of the changes somewhere in the repo itself

**Reply by rdhelms** [reply]
Also, I think in the v3 release notes and documentation above, it would be great to provide information about how to still use 3.x with `experimentalDecorators: true` since some people may want to continue using that (they may have other dependencies that still require it)

**Reply by ruojianll** [reply]
@rdhelms You have enough time which you want to test v3. I will create a changes dir to trace what's news/changes in every future version.

**Reply by rdhelms** [reply]
@ruojianll I just tested our app with `3.0.0-beta.3` and keeping `experimentalDecorators: true` and everything still seems to work great! The functionality seems ready to go as far as I can tell 👍 

I think it would still be helpful to have some good release notes in order for everyone to know that you can still use 3.x with `experimentalDecorators: true` without breaking changes

**Reply by ruojianll** [reply]
@rdhelms Would you test it with `experimentalDecorators: false` in your project? It should be compatible well.

**Reply by rdhelms** [reply]
@ruojianll I'll take a look on Tuesday 👍

Just to clarify - do you mean with `experimentalDecorators: false` and with using `toNative`? Or did you somehow make it work to turn off experimental decorators without needing toNative?

**Reply by ruojianll** [reply]
@rdhelms  `toNative` is a very simple function. So you could test v3 just on you code base, using `toNative` or not is not matter.

**Reply by rdhelms** [reply]
@ruojianll It seems like `3.0.0-beta.3` was working well in my project, although I tried to briefly test it with a fresh Vue 3 app using `create-vue` and I'm getting a `Uncaught SyntaxError: Invalid or unexpected token` error when `experimentalDecorators` is false. Are you able to take a look? Here's a branch with what I was testing:
https://github.com/rdhelms/vue3-migration-reference/tree/vfd-v3-testing

**Comment by kliehm** [comment]
In `vue-property-decorator` v10.rc3 or `vue-class-decorator` v8.rc1 they use `@Options` instead of `@Component`. Would this be an option?
With the current (not next) release of `vue-facing-decorator`, `vue-test-utils` throws an error "Class constructors cannot be invoked without 'new'" when I `mount(MyComponent)`. Could this be related to the `@Component`decorator?

**Reply by ruojianll** [reply]
https://github.com/facing-dev/vue-facing-decorator/issues/54#issuecomment-1500019388
